### DISCUSSION TITLE
How are plugins going to work?

### DISCUSSION THREAD

**Discussion by ryanmrichard** [NWChemEx/NWChemEx#44] [Q&A]
The idea of NWChemEx is that if I want to develop a plugin, I grab PropertyTypes (which grabs all of its dependencies) and then I go about writing my module(s). Importantly my development can be independent of the rest of NWX in the sense that I don't need to modify any other part of NWX or other existing modules. However, right now NWX requires everything to be linked together, which in turn requires modifying the NWChemEx repo so that it knows about all of the plugins that people are writing. We don't want it to work this way.

IMHO the ideal solution would be to make it so that grabbing PropertyTypes would also contain the infrastructure necessary to generate Python bindings for the newly written modules. Then we can rely on Python to do the dynamic loading. Specifically the developer could do something like:

```.py
import nwchemex as nwx # This is the off-the-shelf nwx package
import my_plugin

mm = sde.ModuleManager()
nwx.load_modules(mm)
my_plugin.load_modules(mm)

# Business as usual...
```

The other option I see is to have the NWChemEx repo mess around with dynamic loading. From Python this would look
something like:
```.py
import nwchemex as nwx # Still off-the-shelf

mm = sde.ModuleManager()
nwx.load_modules(mm)
nwx.load_modules_from_shared_library(mm, "/path/to/shared/library.so")

# Business as usual...
```
Namely in NWChemEX we'd have to write a function `load_modules_from_shared_library` which takes care of dynamically loading the library. If you've never messed around with this sort of stuff, I'll just tell you it's a pain and not an attractive option.

So I guess the questions are: is option 1 doable? are there other options? @wadejong I guess the first question is largely targeted at you.

**Comment by wadejong** [comment]
Option 1 will work out of the box.

**Reply by ryanmrichard** [reply]
Is there a time frame for option 1 working? If it already does, can you write up some documentation, or an example, because I'm honestly not sure how to do it.

**Comment by wadejong** [comment]
Not sure what you mean by "...grabbing PropertyTypes would also contain the
infrastructure necessary to generate Python bindings for the newly written
modules."

Things to decide on:

1. Right now we have the Python __init__ centralized in NWChemEx as you
said via "import nwchemex as nwx". We can decentralize this, or we need at
least a base __init__ of the NWChemEx core infrastructure modules (SDE,
etc.).

2. Do we want an __init__ for each module? We can use a simplified version
of the Findcppyy cmake function that generates a short __init__ that
exposes the proper header files with the functions you want to expose and
loads the library. And, we could do this for each module that we want to
expose. I.e., something like MakePythonBindings.cmake, that generates the
__init__ . We would need to take care of dependencies, I.e. PropertyTypes
__init__ needs to import the libchemist __init__ as it is a dependency.

Alternatively, you PropertyTypes __init__ can laod all it's dependencies in
it's import call.

3. We have to decide how to load dependencies, init TA, and what should be
a core set of modules to load. Right now, NWChemEx's Python init loads sde,
libchemist, nwx, property_types and TA directly.

If we want access to integrals in Python, as an example, we could load this
as a plugin or as a core set of Python functions.

4. We can also make the nwx header simpler, by loading each individual
module __init__ . But, somewhere TA needs to be initialized, and some other
pieces loaded.

**Reply by ryanmrichard** [reply]
> Not sure what you mean by "...grabbing PropertyTypes would also contain the
infrastructure necessary to generate Python bindings for the newly written
modules."

The scenario is that a developer wants to write modules for NWX, but keep the code outside of the main NWX module repos (currently defined as Integrals, SCF, MP2, and TAMM-SDE). I primarily seeing this scenario occurring when the developer is maintaining their own code base, but wants to interface with NWX.  Anyways, that developer only needs to make their repo depend on PropertyTypes, as that's what defines all the APIs. At the end of their development cycle they now have a bunch of modules sitting in their repo and only have access to NWX infrastructure found in PropertyTypes and its dependencies. That NWX infrastructure should be set-up so that the developer can do something like:

```.py
import property_types # NWX infrastructure up to property types
import my_modules           # modules the developer wrote

mm = sde.ModuleManager() # in scope from import property_types
my_modules.load_modules(mm)

a_property_type = # Set the property type of the module
key =                      # Key of the module the developer wrote
args =                    # Arguments to the module
result = mm.run_as<a_property_type>(key, *args)
```
As I understand it all the infrastructure for exposing things to Python lives (more-or-less) in the `Findcppyy.cmake` file, which right now exists in the NWChemEx repo and is only accessible in NWChemEx's cmake files.

> 1. Right now we have the Python __init__ centralized in NWChemEx as you
said via "import nwchemex as nwx". We can decentralize this, or we need at
least a base __init__ of the NWChemEx core infrastructure modules (SDE,
etc.).

I feel like each repo starting with Runtime and going all the way up the software stack (so SDE, LibChemist, PropertyTypes, Integrals, etc.) should have its own `__init__`. This also makes it easier to separate concerns as we can handle Python repo by repo.

> Do we want an __init__ for each module? We can use a simplified version
of the Findcppyy cmake function that generates a short __init__ that
exposes the proper header files with the functions you want to expose and
loads the library. And, we could do this for each module that we want to
expose. I.e., something like MakePythonBindings.cmake, that generates the
__init__ . We would need to take care of dependencies, I.e. PropertyTypes
__init__ needs to import the libchemist __init__ as it is a dependency.
>
> Alternatively, you PropertyTypes __init__ can laod all it's dependencies in
it's import call.

If by module you mean repo, then I think so. Each SDE module is on the granularity of a function where the repos are the granularity of a typical Python module. If we move the `Findcppyy.cmake` CMake module (lots of things called modules...) to Runtime, then repos which depend on Runtime should have the functions defined in `Findcppyy.cmake` available in their builds as well.

> 3. We have to decide how to load dependencies, init TA, and what should be
a core set of modules to load. Right now, NWChemEx's Python init loads sde,
libchemist, nwx, property_types and TA directly.
> 
> If we want access to integrals in Python, as an example, we could load this
as a plugin or as a core set of Python functions.

I think that each repo's `__init__` should be responsible for loading its dependencies (I presume Python already does something smart if two Python modules both import the same Python module, but I've never looked into it). This means that initializing MADNESS (and MPI) is going to fall to Runtime, initializing TA is going to fall to libchemist. The user then gets to choose what sets of modules to load in their Python script:

```.py
import integrals
import scf
import mp2

mm = sde.ModuleManager() # in scope from any of the three Python modules
integrals.load_modules(mm) # Now has Integral's modules
scf.load_modules(mm) # Now has Integrals' and SCF's modules
mp2.load_modules(mm) # No has Integrals', SCF's, and MP2's modules
```
The NWChemEx repo is basically just a convenience function for doing this and "cross-linking" them (i.e., as loaded this way SCF's modules still need their integral submodules to be set, whereas MP2's modules still need their integral and SCF sub modules to be set).

**Comment by wadejong** [comment]
Also, as Python spans all modules, where would we be putting such
documentation?

**Reply by ryanmrichard** [reply]
Assuming we move `Findcppyy.cmake` to Runtime, then I think instructions for exposing code to Python via the `Findcppyy.cmake` CMake module should go there. Aside from that, Runtime probably doesn't need any other Python-related documentation. I see the bulk of the Python documentation living in SDE, which is where I think we should have Python tutorials covering:
- writing a module in Python
- writing a property type in Python (if it's possible; it might not be)
- making a module repository (whatever we want to call repos like Integrals, SCF, and MP2)
  - Would probably link to Runtime documentation for convenience
  - Talks about `load_module` function
There's corresponding tutorials for C++ which could just be Pythonized. These tutorials are a bit generic so that they're not coupled to anything downstream.

Beyond the SDE repo everything is just an implementation/application of the SDE framework so I think it can mainly just link back to the SDE documentation. A possible caveat is that the NWChemEx repo is sort of the top-level repo, and it might make sense to revisit some of the tutorials with the actual NWX objects/property types.

@wadejong What do you think?

**Comment by wadejong** [comment]
When I say modules, I mean repos (coarse grain).

On:

import property_types # NWX infrastructure up to property types
import my_modules           # modules the developer wrote

mm = sde.ModuleManager() # in scope from import property_types
my_modules.load_modules(mm)

> So, any external plugin needs to write a load_modules AND needs to have
sde headers. To build a link into NWChemEx they need to do work.

An __init__ for each repo will create a lot of duplication, as each repo
will have to make sure it has all the header paths and headers of their
dependencies. I have to look in cmake, but do we even have a variable there
that tells us for the repo at hand what dependencies it has, and what
dependencies those dependencies have?

Unless we have an NWChemExCore that loads all the fundamentals. I realize I
am arguing for something that is more monolithic, which is probably not the
right way to go. But, realistically, once a repo is married to sde, etc.
there is a core that they depend on. And plugins will end up having similar
dependencies.


I'm a little lost as to what this does:

a_property_type = # Set the property type of the module
key =                      # Key of the module the developer wrote
args =                    # Arguments to the module
result = mm.run_as<a_property_type>(key, *args)

Other than using the NWChemEx runtime, there is no connection to
NWChemEx. What is an explicit example? I can run a forcefield code
this way and get coordinates? I can run an external plugin that gives
me integrals I didn't have that I can then add to my Fock matrix? I am
missing the free standing nature of plugins in connection with
NWChemEx. Maybe a solver? But, that one needs TA to get to the data.

**Reply by ryanmrichard** [reply]
>On:
>
>>import property_types # NWX infrastructure up to property types
>>import my_modules           # modules the developer wrote
>>
>>mm = sde.ModuleManager() # in scope from import property_types
>>my_modules.load_modules(mm)
>
> So, any external plugin needs to write a load_modules AND needs to have
> sde headers. To build a link into NWChemEx they need to do work.

Correct. Every NWX plugin is going to have PropertyTypes, LibChemist, SDE, and Runtime as dependencies (as well as the dependencies of PropertyTypes, LibChemist, SDE, and Runtime). They then need to wrap their code in a module and then they need to make sure that module can be loaded into the SDE. In practice this wrapper code can be pretty minimal:

```.cpp
#include <property_types/energy.hpp>
#include <sde/sde.hpp> // May actually be included via energy.hpp

// Tells NWX how it can call our module
using property_type = PropertyTypes::Energy<double>;

namespace my_repo {

inline load_modules(sde::ModuleManger& mm) {
   
    // Give our code the necessary API
    auto l = [](const libchemist::Molecule& mol, std::size_t deriv) {
       // call my function which computes the energy
    };
    // Tell NWX it can call our module now
    mm.add_module("My Module", sde::make_lambda<property_type>(l));
}
```
of course they can also fully buy into the SDE framework and write a real (as opposed to a lambda) module.

> An __init__ for each repo will create a lot of duplication, as each repo
will have to make sure it has all the header paths and headers of their
dependencies. I have to look in cmake, but do we even have a variable there
that tells us for the repo at hand what dependencies it has, and what
dependencies those dependencies have?

In order to make the C++ library you already have to have all the header paths (and the library paths for a static library), including for dependencies. CMake and the compiler are pretty smart about it though so it's not too bad. The CMake targets are essentially a dependency graph and you can walk the graph to get all this info.

> Unless we have an NWChemExCore that loads all the fundamentals. I realize I
am arguing for something that is more monolithic, which is probably not the
right way to go. But, realistically, once a repo is married to sde, etc.
there is a core that they depend on. And plugins will end up having similar
dependencies.

Right. That core is the stack up to PropertyTypes. They don't have to be wed to anything else.

> I'm a little lost as to what this does:
>
>> a_property_type = # Set the property type of the module
>> key =                      # Key of the module the developer wrote
>> args =                    # Arguments to the module
>> result = mm.run_as<a_property_type>(key, *args)

It's just a normal call to a module, where I explicitly pointed out that the user needs to provide three pieces of information: the property type of the module, the key the module is stored under, and the arguments to the module. More concretely it would look something like:
```.py
J = mm.run_as[nwx.scf.pt.coulomb[double, nwx.scf.orbitals]]("CanJK", mol, occ, bs, bs)
```

> Other than using the NWChemEx runtime, there is no connection to
NWChemEx. What is an explicit example? I can run a forcefield code
this way and get coordinates? I can run an external plugin that gives
me integrals I didn't have that I can then add to my Fock matrix? I am
missing the free standing nature of plugins in connection with
NWChemEx. Maybe a solver? But, that one needs TA to get to the data.

I think part of the confusion lies in how we partition the software stack. To me, the SDE is a generic framework for writing plugin-based software, which will eventually rely on Runtime for runtime information. Anybody can use the SDE to write plugin-based software. The SDE punts on defining what the plugin APIs are leaving it to programmers in the domain to do that. In theory this is where the electronic structure community comes in. If we can all agree on APIs for what it means to compute say an energy, those standard APIs become APIs for the plugins (which we would have to codify into property types). To my knowledge MolSSI hasn't gotten to this level of standardization yet, so we don't have any such standard APIs to work with. Thus we wrote PropertyTypes to fill this role, using the classes defined in LibChemist (in theory LibChemist shouldn't really depend on SDE like it does). Any software that uses the APIs defined in PropertyTypes is interoperable with any other software that uses those APIs, even if that software doesn't use the SDE for plugin management.  I'd argue everything up to and including PropertyTypes should be considered the framework for NWX, not NWX. To me NWX is the collection of modules that we've written to be used with that framework (so Integrals, SCF, MP2, etc.). 

So with that all said. A "plugin" is really just a module or set of modules that are separate from the canonical set of NWX modules. As far as an explicit example I'm (very slowly) working on the [GhostFragment](https://github.com/rmrresearch/GhostFragment) plugin, which will do basis-set superposition error corrections and the many-body expansion. From my perspective, GhostFragment is a standalone piece of software, which is complimentary to NWX, but not part of NWX. To do its thing, GhostFragment needs to call things that compute energies. Again, lacking community consensus, GhostFragment adopts the APIs defined in PropertyTypes. So if another package uses those APIs, or is wrapped so that it has those APIs, GhostFragment can drive it. In particular for my current purposes, if I wrap Psi4 so that it's callable via the energy API, I can use GhostFragment with Psi4.

Does that clear it up?

**Comment by wadejong** [comment]
All test cases in all the repos can be Pythonized, the SDE one is already
there.

> writing a property type in Python (if it's possible; it might not be)

Not sure what this buys you.

**Reply by ryanmrichard** [reply]
> Not sure what this buys you.

Rapid prototyping. Plus if you're writing a module in Python and you need a new property type it allows you to keep your package purely in Python.

**Comment by wadejong** [comment]
On:

import integrals
import scf
import mp2

mm = sde.ModuleManager() # in scope from any of the three Python modules
integrals.load_modules(mm) # Now has Integral's modules
scf.load_modules(mm) # Now has Integrals' and SCF's modules
mp2.load_modules(mm) # No has Integrals', SCF's, and MP2's modules


Do you really want a user to do an "import integrals" before "import scf"?
I guess I need to do this at the C++ level too? I.e., scf.load_modules(mm)
does not imply an automatic integrals.load_modules(mm)?

Now a user needs to know which modules/repos depend on each other.


Maybe it's not such a big deal, given most end-users can use a cross-linked
version.

**Reply by ryanmrichard** [reply]
>Do you really want a user to do an "import integrals" before "import scf"?
I guess I need to do this at the C++ level too? I.e., scf.load_modules(mm)
does not imply an automatic integrals.load_modules(mm)?
>
> Now a user needs to know which modules/repos depend on each other.
>
>
> Maybe it's not such a big deal, given most end-users can use a cross-linked
version.

SCF and MP2 are decoupled from Integrals and each other. They are only coupled to the APIs defined in property types. In theory this allows someone to come along and grab say the SCF modules, and to use the SCF modules with their own integrals routines, i.e. they don't have to use the modules defined in Integrals. Of course, at some point SCF is going to need integrals modules. In our stack it gets them in the NWChemEx repo. We put all of the cross-linking off until the very end. The idea is that this potentially benefits other developers (who can now grab subsets of modules) and it is transparent to the end-user who grabs the NWChemEx repo (which is completely cross-linked and is ready to go). Admittedly the process of cross-linking modules is a pain and I'd be open to suggestions on how to make it easier without compromising the separation of concerns we currently have.

**Comment by wadejong** [comment]
That definitely clears things up.

I would then suggest we have a defined Core to import that includes
PropertyTypes,
LibChemist, SDE, and Runtime and it's dependencies. As we will always need
these, that would be the first to import in Python.

**Reply by ryanmrichard** [reply]
What I'd like to do is call what you're calling "Core" the SDE and change what we're currently calling SDE to something like "PluginPlay". Since PropertyTypes is equivalent to what you're calling "Core" I'd suggest renaming it to SDE. We also may want to take this time to change the abbreviation to something other than SDE, maybe SimDE (play on single instruction-multiple data). I'm sure there's going to be wide-spread grumblings about renaming, but I think it clears things up...

**Comment by wadejong** [comment]
I get that. For developers, the current separation will be fine, as long as
we document it. When it comes to end-users (running NWChemEx) we can build
a high-level NWChemEx with all the necessary components exposed...though
they probably would use even higher level Python wrappers that hide more
aspects.

**Comment by wadejong** [comment]
Fine by me. We really should have a core of libraries that everything is
build upon, which also be the core to load into Python.

**Reply by ryanmrichard** [reply]
@wadejong is it possible that in the future you can respond on GitHub? When you respond via email it keeps making new "answers" instead of replies. This makes the conversation hard to parse as your replies aren't tied to what they're referencing. It probably doesn't matter if you and I are the only two participating, but if someone else gets involved or tries to read it I can imagine it would be difficult.

**Reply by wadejong** [reply]
Sure.
